pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java

```diff
@@ -244,6 +258,66 @@ public void start() throws Exception {
+                + "-" + instanceConfig.getFunctionAuthenticationSpec().getData();
+    }
+
+    private void createServiceAccount() throws ApiException, InterruptedException {
+
+        String fqfn = FunctionDetailsUtils.getFullyQualifiedName(instanceConfig.getFunctionDetails());
+        Function.FunctionAuthenticationSpec authenticationSpec = instanceConfig.getFunctionAuthenticationSpec();
+        String serviceAccountName = generateServiceAccount(instanceConfig);
+        V1ServiceAccount serviceAccount = new V1ServiceAccount()
+                .metadata(
+                        new V1ObjectMeta()
+                                .name(serviceAccountName)
+                                .namespace(jobNamespace));
+
+        // configure service account for auth data if necessary
+        functionAuthDataCacheProvider.configureAuthDataKubernetesServiceAccount(authenticationSpec, serviceAccount);
+
+        log.info("Creating service account with the following spec to k8 {} for function {}", appsClient.getApiClient().getJSON().serialize(serviceAccount), fqfn);
+
+        RuntimeUtils.Actions.Action createServiceAccount = RuntimeUtils.Actions.Action.builder()
+                .actionName(String.format("Creating service account %s for function %s", serviceAccountName, fqfn))
+                .numRetries(KubernetesRuntimeFactory.NUM_RETRIES)
+                .sleepBetweenInvocationsMs(KubernetesRuntimeFactory.SLEEP_BETWEEN_RETRIES_MS)
+                .supplier(() -> {
+                    try {
+                        coreClient.createNamespacedServiceAccount(jobNamespace, serviceAccount, "true");
+                     } catch (ApiException e) {
+                        // already exists
+                        if (e.getCode() == HTTP_CONFLICT) {
+                            log.warn("Service account {} already present for function {}", serviceAccountName, fqfn);
+                            return RuntimeUtils.Actions.ActionResult.builder().success(true).build();
+                        }
+
+                        String errorMsg = e.getResponseBody() != null ? e.getResponseBody() : e.getMessage();
+                        return RuntimeUtils.Actions.ActionResult.builder()
+                                .success(false)
+                                .errorMsg(errorMsg)
+                                .build();
+                    }
+
+                    return RuntimeUtils.Actions.ActionResult.builder().success(true).build();
+                })
+                .build();
+
+        AtomicBoolean success = new AtomicBoolean(false);
+        RuntimeUtils.Actions.newBuilder()
+                .addAction(createServiceAccount.toBuilder()
+                        .onSuccess(() -> success.set(true))
+                        .build())
+                .run();
+
+        if (!success.get()) {
+            throw new RuntimeException(String.format("Failed to create service account %s for function %s", serviceAccountName, fqfn));
```

Comment:
logging before throwing runtime exception

Reply:
it will be logged else where

Reply:
> it will be logged else where

sorry, at where? if you don't catch RuntimeException, isn't the context lost?

Reply:
its caught and logged here:

https://github.com/apache/pulsar/pull/3735/files#diff-c2c5e0719db635b36beb4cd6d816abdeL226